core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -810,7 +810,7 @@ private[spark] object Utils extends Logging {
    * Get the local machine's URI.
    */
   def localHostNameForURI(): String = {
-    customHostname.getOrElse(InetAddresses.toUriString(localIpAddress))
+    sys.env.get("SPARK_PUBLIC_DNS").getOrElse( customHostname.getOrElse(InetAddresses.toUriString(localIpAddress)) )
```

Comment:
sorry, not got much insight there, except failing fast on "" is good. It should really be the job of the code downstream to recognise and react to the usual config problems of trying to listen on an invalid hostname or port 0